### PR TITLE
Initialize an `IamDataFrame` from `pd.DataFrame` with formatting specs

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

This PR allows initialising an `IamDataFrame` from a file or `pd.DataFrame` and filling out missing required columns efficiently.

See the unit tests in `tests/test_cast_to_iamc.py` for possible use cases.

### CODE REVIEW

**Comment by danielhuppmann** [comment]
@znicholls, this PR tries to make it easier to initialize an `IamDataFrame` which doesn't quite have the right format, for example when you have a `pd.DataFrame` in a `scen`-file format with values in multiple columns and the column name being the identifier. See also the discussion here https://github.com/IAMconsortium/pyam/pull/193#issuecomment-464634324.

Can you check whether that works with your thinking for the `ScmDataFrame`?

**Comment by znicholls** [comment]
will leave final review and merge to @gidden nice tweak @danielhuppmann

File: pyam/core.py [hunk 1]
```diff
@@ -1501,13 +1500,13 @@ def compare(left, right, left_label='left', right_label='right',
 
 def concat(dfs):
     """Concatenate a series of `pyam.IamDataFrame`-like objects together"""
-    if not hasattr(dfs, '__iter__'):
+    if isstr(dfs) or not hasattr(dfs, '__iter__'):
         raise TypeError('Input data must be iterable (e.g., list or tuple)')
```
Comment (gidden) [inline pyam/core.py]: can you add here "non-string iterable" just for consistency..

File: pyam/utils.py [hunk 2]
```diff
@@ -194,18 +235,18 @@ def format_data(df):
...
-def style_df(df, style='heatmap'):
-    if style == 'highlight_not_max':
-        return df.style.apply(lambda s: ['' if v else 'background-color: yellow' for v in s == s.max()])
-    if style == 'heatmap':
-        cm = sns.light_palette("green", as_cmap=True)
-        return df.style.background_gradient(cmap=cm)
+
+def sort_data(data, cols):
+    """Sort `data` rows and order columns"""
+    return data.sort_values(cols)[cols + ['value']].reset_index(drop=True)
```
Comment (gidden) [inline pyam/utils.py]: what happens if "value" is already in `cols`? not sure if this is a corner case, but could do `list(set(cols + ['value']))`

**Comment by gidden** [comment]
ok, small changes requested. looks great!

File: pyam/utils.py (hunk 2 above)
Comment (danielhuppmann) [reply pyam/utils.py]: Usig `list(set(cols + ['value']))` would defeat the purpose of having a consistent ordering of columns, which makes the unit tests a lot more straightforward by using `pd.testing.assert_frame_equal()`.

I tried migrating that function to `core.py`, but then this function can't be used in the unit tests, which is opening up another can of worms.

File: pyam/utils.py [hunk 3]
```diff
@@ -395,3 +436,16 @@ def to_int(x, index=False):
         return x
     else:
         return _x
+
+
+def concat_with_pipe(x, cols=None):
+    """Concatenate a `pd.Series` separated by `|`, drop `None` or `np.nan`"""
+    return '|'.join([x[i] for i in cols or x.index
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply pyam/utils.py]: done

File: pyam/core.py (hunk 1 above)
Comment (danielhuppmann) [reply pyam/core.py]: done

**Comment by gidden** [comment]
thanks @danielhuppmann !

**Pull request merged**